src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/Vector256.cs

```diff
@@ -62,10 +65,70 @@ public static bool IsHardwareAccelerated
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static Vector256<T> Abs<T>(Vector256<T> vector)
         {
-            return Create(
-                Vector128.Abs(vector._lower),
-                Vector128.Abs(vector._upper)
-            );
+            if ((typeof(T) == typeof(byte))
+             || (typeof(T) == typeof(ushort))
+             || (typeof(T) == typeof(uint))
+             || (typeof(T) == typeof(ulong))
+             || (typeof(T) == typeof(nuint)))
+            {
+                return vector;
+            }
+            else if (Avx.IsSupported)
+            {
+                return XarchImpl(vector);
+            }
+            return SoftwareImpl(vector);
+
+            [CompExactlyDependsOn(typeof(Avx))]
+            [CompExactlyDependsOn(typeof(Avx2))]
+            [CompExactlyDependsOn(typeof(Avx512F.VL))]
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static Vector256<T> XarchImpl(Vector256<T> vector)
+            {
+                if (sizeof(T) == sizeof(float))
+                {
+                    return Avx.AndNot(Vector256.Create(-0.0f), vector.AsSingle()).As<float, T>();
+                }
+                else if (sizeof(T) == sizeof(double))
+                {
+                    return Avx.AndNot(Vector256.Create(-0.0), vector.AsDouble()).As<double, T>();
+                }
+                else if (Avx2.IsSupported)
+                {
+                    if (sizeof(T) == 1)
+                    {
+                        return Avx2.Abs(vector.AsSByte()).As<byte, T>();
+                    }
+                    else if (sizeof(T) == 2)
+                    {
+                        return Avx2.Abs(vector.AsInt16()).As<ushort, T>();
+                    }
+                    else if (sizeof(T) == 4)
+                    {
+                        return Avx2.Abs(vector.AsInt32()).As<uint, T>();
+                    }
+                    else if (sizeof(T) == 8)
+                    {
+                        if (Avx512F.VL.IsSupported)
```

Comment:
They're only dynamic feature checks for the same encoding levels.

That is, if you target the baseline (`x86-64-v1`) then `SSE3-SSE4.2` become dynamic checks while `Avx+` is static false (due to requiring VEX encoding).
If you target `x86-64-v2` then there should be no dynamic checks.
If you target `x86-64-v3` then there should be no dynamic checks (`Avx512F+` requires EVEX encoding).
If you target `x86-64-v4` then you would get dynamic checks for `Avx512Vbmi`

You could also target individual ISAs, but that's not typically recommended. There are then also some one off ISAs not in the normal hierarchy, such as `Aes` and `Pclmulqdq`, but those are exceptionally rare. The same would be true of ISAs that are not part of a target baseline but which support an encoding, such as `AvxVnni`.

So given the current code here, I expect that NAOT will do just fine. There may be some concern for other areas, but that's likely something we want a feature or attribute to help handle anyways given the same concern applies to user code that is writing small helper methods that do per architecture dispatch.

For S.P.Corelib in particular, it's also worth noting that the trimmer is going to remove the `if (AdvSimd.IsSupported)` and similar paths when targeting xarch, so in many cases we're going to get down to just the 4 constant size checks.

Reply:
I agree that we are lucky that this works given other constraints for typical configurations. It does not work well in general.

Reply:
> It does not work well in general.

Agreed. I think this is an area we're going to need to think about more in depth for NAOT when it comes time.

In many cases, the dynamic checks aren't worthwhile unless you have some form of code cloning or are executing more complex logic. While the pattern we tend to encourage for users is to write small helpers that isolate the functionality instead (as a way of improving readability and maintainability).

Perhaps that could be some `bool IsSupportedDynamically { get; }` (or `IsSupportedOpportunistically`?) that allows the developer to be explicit they would like a dynamic check instead of an exact check.